### PR TITLE
leveldb cleaner

### PR DESCRIPTION
**Pull request by joshblum**
Implements a simple db cleaner on leveldb. get/puts are tracked in an in-memory  LRU (dropped the disk backing for simplicity, we can bring it in if we feel this performs poorly), unused keys are cleaned periodically basic on a configuration of desired size. we cycle through the entire level db over the course of our cleaning based on an in-memory pointer to our location in the db so we don't unfairly start purging from the beginning at each run.

### CODE REVIEW

**Comment by mmaxim** [comment]
I'm sure you checked this, but LevelDB doesn't have any kind of stats like this builtin?

**Comment by joshblum** [comment]
nothing that i could find in the docs/reading the code

File: go/libkb/dbcleaner.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+func (c *levelDbCleaner) clearCache() {
+	c.cache.Purge()
+}
+
+func (c *levelDbCleaner) clean(ctx context.Context) (err error) {
+	c.Lock()
+	defer c.Unlock()
+	if !c.forceClean && c.G().GetClock().Now().Sub(c.lastRun) < c.config.CleanInterval {
```
Comment (joshblum) [inline go/libkb/dbcleaner.go]: only run i background active mode on mobile

**Comment by joshblum** [comment]
@mmaxim updated ptal
updates:
- monitor the app state to attempt mobile cleaning only when in the BACKGROUNDACTIVE state.
- require a minimum amount of values in the cache before running a clean
- whitelist keys we don't want/need cleaned, introduces a new 'table' to store these values in
- cli/gui trigger options

**Comment by joshblum** [comment]
cc @keybase/react-hackers for admin only trigger

File: go/libkb/dbcleaner.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+	if !force && dbSize < c.config.MaxSize {
+		return nil
+	}
+
+	var totalNumPurged, numPurged int
+	for i := 0; i < 100; i++ {
+		select {
+		case <-c.cancelCh:
+			c.log(ctx, "aborting clean, %d runs, canceled", i)
+			break
```
Comment (patrickxb) [inline go/libkb/dbcleaner.go]: I don't think these `break` statements are doing what you want.  You would need a label to break out of the `for` loop here.

see the 3rd example here:

https://golang.org/doc/effective_go.html#switch

File: go/libkb/leveldb_test.go [hunk 3]
```diff
@@ -94,35 +97,66 @@ func TestLevelDb(t *testing.T) {
...
+
+				// this key will not be deleted since it is in the permanent
+				// table.
+				require.True(t, IsPermDbKey(DBDiskLRUEntries))
+				permKey := DbKey{Key: "test-key", Typ: DBDiskLRUEntries}
+				err = db.Put(permKey, nil, v)
+				require.NoError(t, err)
+
+				// cleaner will not clean the key since it was recently used
+				err = db.cleaner.clean(context.TODO(), true /* force */)
```
Comment (patrickxb) [inline go/libkb/leveldb_test.go]: check this error.

File: go/libkb/leveldb_test.go [hunk 4]
```diff
@@ -94,35 +97,66 @@ func TestLevelDb(t *testing.T) {
...
+				_, found, err := db.Get(key)
+				require.NoError(t, err)
+				require.True(t, found)
+				_, found, err = db.Get(permKey)
+				require.NoError(t, err)
+				require.True(t, found)
+
+				db.cleaner.clearCache()
+				err = db.cleaner.clean(context.TODO(), true /* force */)
```
Comment (patrickxb) [inline go/libkb/leveldb_test.go]: check this error.

File: go/libkb/dbcleaner.go (hunk 2 above)
Comment (joshblum) [reply go/libkb/dbcleaner.go]: ah great catch

File: go/libkb/leveldb_test.go [hunk 5]
```diff
@@ -229,20 +249,16 @@ func TestLevelDb(t *testing.T) {
 					defer wg.Done()
 
 					tr, err := db.OpenTransaction()
-					if err != nil {
-						fmt.Println(err)
-						t.Fatal(err)
-					}
+					require.NoError(t, err)
```
Comment (patrickxb) [inline go/libkb/leveldb_test.go]: the existing test is also broken, but not supposed to call t.Fatal inside a goroutine.  (and `require` package uses `t.FailNow` which is same)...   `if err != nil { t.Error(err) }` would be fine.

Comment (joshblum) [reply go/libkb/leveldb_test.go]: fixed in https://github.com/keybase/client/pull/16263

**Pull request merged**